File: labs/typescript/example_api_version_check.ts
/**
 * ApiVersionResource check (TypeScript) — mirrors example_api_version_check.py.
 * No credentials required.
 */

import { mkdirSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { getInternetmarkeBaseUrl, loadInternetmarkeConfig } from "@gruncellka/porto-sdk";
import { loadLabEnv } from "../../lib/typescript/load_env.ts";

const __dirname = dirname(fileURLToPath(import.meta.url));
loadLabEnv();

function saveResult(payload: Record<string, unknown>): string {
  const outDir = process.env.OBSERVER_RUN_DIR
    ? join(process.env.OBSERVER_RUN_DIR, "api")
    : join(__dirname, "artifacts", "api");
  mkdirSync(outDir, { recursive: true });
  const ts = new Date().toISOString().replace(/[:.]/g, "-").slice(0, 19);
  const outPath = join(outDir, `api_version_${ts}.json`);
  writeFileSync(outPath, `${JSON.stringify(payload, null, 2)}\n`, "utf8");
  return outPath;
}

async function main(): Promise<number> {
  const im = loadInternetmarkeConfig("deutschepost");
  const baseUrl = getInternetmarkeBaseUrl(im);
  const versionUrl = `${baseUrl}/`;

  console.log("ApiVersionResource check (TS)");
  console.log(`GET ${versionUrl}`);

  let ok = true;
  let result: Record<string, unknown>;
  try {
    const resp = await fetch(versionUrl);
    const text = await resp.text();
    result = { status: resp.status, ok: resp.ok, body_preview: text.slice(0, 2000) };
    ok = resp.ok;
  } catch (error) {
    ok = false;
    result = { error: String(error) };
  }

  const path = saveResult({
    ts: new Date().toISOString(),
    status: ok ? "ok" : "error",
    response_json: result,
  });

  console.log(ok ? "API available" : "API check failed");
  console.log(`Saved: ${path}`);
  return ok ? 0 : 1;
}

main().then((code) => process.exit(code));

File: labs/python/example_api_version_check.py
"""
ApiVersionResource check — uses Porto SDK health check. No credentials required.
"""

from __future__ import annotations

import asyncio
import json
import os
import sys
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

_REPO_ROOT = Path(__file__).resolve().parents[2]
if str(_REPO_ROOT) not in sys.path:
    sys.path.insert(0, str(_REPO_ROOT))

from labs.lib.python.internetmarke_health import health  # noqa: E402
from labs.lib.python.load_env import load_lab_env  # noqa: E402

load_lab_env()

from porto_sdk.adapters.deutschepost.internetmarke.bootstrap import (  # noqa: E402
    get_internetmarke_base_url,
    load_internetmarke_config,
)


def save_result(payload: dict[str, Any]) -> Path:
    observer_run_dir = os.getenv("OBSERVER_RUN_DIR")
    if observer_run_dir:
        out_dir = Path(observer_run_dir) / "api"
    else:
        out_dir = Path(__file__).parent / "artifacts" / "api"
    out_dir.mkdir(parents=True, exist_ok=True)

    ts = datetime.now(UTC).strftime("%Y%m%d-%H%M%S")
    out_path = out_dir / f"api_version_{ts}.json"
    out_path.write_text(json.dumps(payload, indent=2) + "\n", encoding="utf-8")
    return out_path


async def main() -> int:
    im = load_internetmarke_config("deutschepost")
    base_url = get_internetmarke_base_url(im)
    version_url = f"{base_url.rstrip('/')}/"

    print("ApiVersionResource check (lab health probe)")
    print("=" * 40)
    print(f"Base URL: {base_url}")
    print(f"Endpoint: GET {version_url}")
    print()

    try:
        result = await health(base_url)
        ok = True
    except Exception as exc:  # noqa: BLE001
        ok = False
        result = {"error": str(exc), "type": type(exc).__name__}

    path = save_result(
        {
            "ts": datetime.now(UTC).isoformat(),
            "status": "ok" if ok else "error",
            "response_json": result,
        }
    )

    if ok:
        print("Result: API available")
        for key, value in result.items():
            print(f"  {key}: {value}")
    else:
        print("Result: API check failed")
        print(f"  error: {result.get('error', result)}")

    print(f"\nSaved: {path}")
    return 0 if ok else 1


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main()))
